capsules/src/nrf51822_serialization.rs

```diff
@@ -182,17 +182,66 @@ impl SyscallDriver for Nrf51822Serialization<'_> {
             }
             // Receive from the nRF51822
             2 => {
-                self.rx_buffer.take().map_or(CommandReturn::failure(ErrorCode::RESERVE), |buffer| {
-                    let len = arg1;
-                    if len > buffer.len() {
-                        CommandReturn::failure(ErrorCode::SIZE)
-                    } else {
-                        // Set this as the active app for the receive callback
-                        self.active_app.set(appid);
-                        let _ = self.uart.receive_automatic(buffer, len, 250);
-                        CommandReturn::success_u32(len as u32)
-                    }
-                })
+                let len = arg1;
+
+                // We only allow one app to use the NRF serialization capsule
+                // (old legacy code, and a difficult thing to virtualize).
+                // However, we would like to support restarting/updating apps.
+                // But we don't want to allow a simultaneous app to disrupt the
+                // app that got to the BLE serialization first. So we have to
+                // find a compromise.
+                //
+                // We handle this by checking if the current active app still
+                // exists. If it does, we leave it alone. Otherwise, we replace
+                // it.
+                self.active_app.map_or_else(
+                    || {
+                        // The app is not set, handle this for the normal case.
+                        self.rx_buffer
+                            .take()
+                            .map_or(CommandReturn::failure(ErrorCode::RESERVE), |buffer| {
+                                if len > buffer.len() {
+                                    CommandReturn::failure(ErrorCode::SIZE)
+                                } else {
+                                    // Set this as the active app for the
+                                    // receive callback.
+                                    self.active_app.set(appid);
+                                    let _ = self.uart.receive_automatic(buffer, len, 250);
+                                    CommandReturn::success_u32(len as u32)
+                                }
+                            })
+                    },
+                    |appid| {
+                        // The app is set, check if it still exists.
+                        if let Err(_err) = self.apps.enter(*appid, |_, _| {}) {
```

Comment:
```suggestion
                        if let Err(kernel::process::Error::NoSuchApp) = self.apps.enter(*appid, |_, _| {}) {
```

To protect against future changes to grant where entering the grant might fail for other reasons than `NoSuchApp` / `InactiveApp`